File: 2021/07/27/balanced_binary_tree.py
"""
https://leetcode.com/problems/balanced-binary-tree/
Author: Andrew Jarombek
Date: 7/27/2021
"""

# You are not mean, not even slightly.  I just wish you could trust me to be there for you again if you wanted me to be.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def is_balanced(root: TreeNode) -> bool:
    balanced, height = check_balanced(root)
    return balanced


def check_balanced(root: TreeNode) -> (bool, int):
    if root is None:
        return True, 0
    else:
        left_balanced, left_height = check_balanced(root.left)
        right_balanced, right_height = check_balanced(root.right)

        if not left_balanced or not right_balanced:
            return False, 0
        elif abs(left_height - right_height) <= 1:
            return True, 1 + max(left_height, right_height)
        else:
            return False, 0


if __name__ == '__main__':
    assert is_balanced(TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(2, TreeNode(5), TreeNode(4))))
    assert not is_balanced(TreeNode(1, TreeNode(2, TreeNode(5), TreeNode(4)), None))

File: 2022/11/12/cpp/balancedBinaryTree.cpp
// https://leetcode.com/problems/balanced-binary-tree/
// Author: Andrew Jarombek
// Date: 11/12/2022

#include <queue>
#include <cassert>
#include <iostream>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

bool isBalanced(TreeNode* root) {
    if (root == nullptr)
        return true;

    int maxDepth = -1;
    int minDepth = -1;

    queue<TreeNode*> nodes;
    nodes.push(root);

    while (!nodes.empty()) {
        maxDepth++;
        int length = nodes.size();

        for (int i = 0; i < length; i++) {
            TreeNode* node = nodes.front();
            nodes.pop();

            if (node->right == nullptr || node->left == nullptr) {
                if (minDepth == -1)
                    minDepth = maxDepth;

                if (maxDepth - minDepth > 1) {
                    cout << maxDepth << "\n";
                    cout << minDepth << "\n";
                    return false;
                }
            }

            if (node->left != nullptr) {
                nodes.push(node->left);
            }

            if (node->right != nullptr) {
                nodes.push(node->right);
            }
        }
    }

    cout << maxDepth << "\n";
    cout << minDepth << "\n";
    return maxDepth - minDepth <= 1;
}

int main() {
    auto* tree = new TreeNode(1, new TreeNode(2), new TreeNode(3));
    assert(isBalanced(tree));
}